src/librustc_privacy/lib.rs

```diff
@@ -922,33 +928,38 @@ impl<'a, 'tcx: 'a> SearchInterfaceForPrivateItemsVisitor<'a, 'tcx> {
-            if let Some(node_id) = self.tcx.hir.as_local_node_id(impl_trait_ref.def_id) {
-                let item = self.tcx.hir.expect_item(node_id);
-                let vis = ty::Visibility::from_hir(&item.vis, node_id, self.tcx);
+            self.visit_trait_ref(impl_trait_ref);
+        }
+        self
+    }
 
-                if !vis.is_at_least(self.min_visibility, self.tcx) {
-                    self.min_visibility = vis;
-                }
-                if !vis.is_at_least(self.required_visibility, self.tcx) {
-                    if self.has_pub_restricted || self.has_old_errors {
-                        struct_span_err!(self.tcx.sess, self.span, E0445,
-                                         "private trait `{}` in public interface", impl_trait_ref)
-                            .span_label(self.span, format!(
-                                        "private trait can't be public"))
-                            .emit();
-                    } else {
-                        self.tcx.sess.add_lint(lint::builtin::PRIVATE_IN_PUBLIC,
-                                               node_id,
-                                               self.span,
-                                               format!("private trait `{}` in public \
-                                                        interface (error E0445)", impl_trait_ref));
-                    }
+
+    fn visit_trait_ref(&mut self, trait_ref: ty::TraitRef<'tcx>) -> bool {
+        // Non-local means public (private items can't leave their crate, modulo bugs)
+        if let Some(node_id) = self.tcx.hir.as_local_node_id(trait_ref.def_id) {
+            let item = self.tcx.hir.expect_item(node_id);
+            let vis = ty::Visibility::from_hir(&item.vis, node_id, self.tcx);
+            if !vis.is_at_least(self.min_visibility, self.tcx) {
+                self.min_visibility = vis;
+            }
+            if !vis.is_at_least(self.required_visibility, self.tcx) {
+                if self.has_pub_restricted || self.has_old_errors {
+                    struct_span_err!(self.tcx.sess, self.span, E0445,
+                                     "private trait `{}` in public interface", trait_ref)
+                        .span_label(self.span, format!(
+                                    "private trait can't be public"))
+                        .emit();
+                } else {
+                    self.tcx.sess.add_lint(lint::builtin::PRIVATE_IN_PUBLIC,
+                                           node_id,
+                                           self.span,
+                                           format!("private trait `{}` in public \
+                                                    interface (error E0445)", trait_ref));
                 }
             }
-            impl_trait_ref.super_visit_with(self);
         }
-        self
+
+        trait_ref.super_visit_with(self)
```

Comment:
You should move this to `impl_trait_ref` to avoid recursing twice from `predicates` - can you also rename the method to `check_trait_ref`? That seems more appropriate now.